File: pandas/general_query_06.py
#!/usr/bin/python3

"""General query 6

Student Mobility

Javier Fernández-Bravo Peñuela
Universitat Politècnica de València

Destination country preferred by each country's students.

pandas - general_query_06.py
"""

from pandas_job.pandas_task import compute_task
from domain.student import Student, MobilityType

from sys import argv

def students_favorite_destination_country(students_df):
    students_df['HOST_COUNTRY'] = students_df[['MOBILITYTYPE','COUNTRYOFHOSTINSTITUTION','COUNTRYOFWORKPLACEMENT']]\
        .apply(host_country_udf, axis=1)
    students_df = students_df.dropna(subset=['HOST_COUNTRY'])
    students_df = students_df.groupby(['COUNTRYOFHOMEINSTITUTION', 'HOST_COUNTRY'], as_index=False, sort=False).size()
    students_df = students_df[students_df['size'] == students_df.groupby(['COUNTRYOFHOMEINSTITUTION'])['size'].transform('max')]
    students_favorite_destination_country_df = students_df.sort_values('COUNTRYOFHOMEINSTITUTION')
    return students_favorite_destination_country_df

def host_country_udf(row):
    host_country = ''
    if row['MOBILITYTYPE'] in [MobilityType.STUDY, MobilityType.PLACEMENT]:
        if row['MOBILITYTYPE'] is MobilityType.STUDY:
            host_country = row['COUNTRYOFHOSTINSTITUTION']
        elif row['MOBILITYTYPE'] is MobilityType.PLACEMENT:
            host_country = row['COUNTRYOFWORKPLACEMENT']
    return host_country.rstrip() if host_country else None

# python3 general_query_06.py ../data/Student_Mobility.csv [out/pandas06-output]
if __name__ == "__main__":
    compute_task(students_favorite_destination_country, Student, argv[1:])

File: spark/src/main/java/systematicchaos/studentmobility/GeneralQuery06.java
/**
 * Student Mobility
 * 
 * Javier Fernández-Bravo Peñuela
 * Universitat Politècnica de València
 * 
 * General query 6: Destination country preferred by each country's students.
 * 
 * spark - GeneralQuery06.java
 */

package systematicchaos.studentmobility;

import java.util.List;
import java.util.stream.Collectors;

import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.functions;
import org.apache.spark.sql.api.java.UDF3;
import org.apache.spark.sql.types.DataTypes;

import systematicchaos.studentmobility.domain.Student;
import systematicchaos.studentmobility.spark.SparkJob;
import systematicchaos.studentmobility.spark.SparkJobLauncher;
import systematicchaos.studentmobility.spark.SparkTask;
import systematicchaos.studentmobility.util.KeyCountPair;
import systematicchaos.studentmobility.util.StudentBuilder;

import org.apache.spark.sql.expressions.Window;
import org.apache.spark.sql.expressions.WindowSpec;

public class GeneralQuery06 {

	public static class StudentsFavoriteDestinationCountry implements SparkTask<Integer> {
		
		@Override
		public Iterable<KeyCountPair<Integer>> computeSparkTask(String[] inputFiles) {
			SparkJob<Student> job = new SparkJob<>(GeneralQuery06.class.getName(),
					Student.class, new StudentBuilder());
			List<KeyCountPair<Integer>> destinationCount = null;
			
			try {
				Dataset<Student> students = job.readDataset(inputFiles);
				
				students.sqlContext().udf().register("HOST_COUNTRY",
						(UDF3<String, String, String, String>)(mobilityType, countryHostInstitution, countryWorkPlacement) -> {
							String hostCountry = null;
							if (mobilityType != null && !mobilityType.isEmpty()) {
								if (Student.MobilityType.STUDY.getValue() == mobilityType.charAt(0)) {
									hostCountry = countryHostInstitution;
								} else if (Student.MobilityType.PLACEMENT.getValue() == mobilityType.charAt(0)) {
									hostCountry = countryWorkPlacement;
								}
							}
							return hostCountry;
						}, DataTypes.StringType);
				
				Column hostCountryCol = new Column("hostCountry");
				Dataset<Row> hostCountryStudents = students.withColumn("hostCountry", functions.callUDF("HOST_COUNTRY",
						functions.col("mobilityType"), functions.col("countryOfHostInstitution"), functions.col("countryOfWorkPlacement")))
						.filter(hostCountryCol.notEqual(""))
						.groupBy(new Column("countryOfHomeInstitution"), hostCountryCol).count();
				WindowSpec homeCountryWindow = Window.partitionBy("countryOfHomeInstitution");
				hostCountryStudents = hostCountryStudents.withColumn("maxCount", functions.max(functions.col("count")).over(homeCountryWindow))
						.where(functions.col("count").equalTo(functions.col("maxCount")))
						.drop("maxCount")
						.sort("countryOfHomeInstitution");
				
				int nrows = (int) hostCountryStudents.count();
				destinationCount = hostCountryStudents.takeAsList(nrows).stream()
						.map((s) -> new KeyCountPair<>(String.format("%s\t%s", s.getString(0), s.getString(1)), (int)s.getLong(2)))
						.collect(Collectors.toList());
			} catch (Exception e) {
				System.err.println(e.getClass().getName() + ": " + e.getMessage());
			} finally {
				job.close();
			}
			
			return destinationCount;
		}
	}
	
	// spark-submit --class systematicchaos.studentmobility.GeneralQuery06 --master local[*] target/spark-1.0.0.jar ../data/Student_Mobility.csv [out/spark06-output]
	public static void main(String[] args) throws Exception {
		SparkJobLauncher.launchSparkJob(new StudentsFavoriteDestinationCountry(), args);
	}
}
